Amazon.QLDB.Driver/session/AsyncSessionPool.cs

```diff
@@ -0,0 +1,175 @@
+﻿/*
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"). You may not use this file except in compliance with
+ * the License. A copy of the License is located at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * or in the "license" file accompanying this file. This file is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR
+ * CONDITIONS OF ANY KIND, either express or implied. See the License for the specific language governing permissions
+ * and limitations under the License.
+ */
+
+namespace Amazon.QLDB.Driver
+{
+    using System;
+    using System.Collections.Concurrent;
+    using System.Threading;
+    using System.Threading.Tasks;
+    using Microsoft.Extensions.Logging;
+
+    /// <summary>
+    /// The implementation of the session pool.
+    /// </summary>
+    internal class AsyncSessionPool : BaseSessionPool
+    {
+        private readonly BlockingCollection<AsyncQldbSession> sessionPool;
+        private new readonly SemaphoreSlim poolPermits;
+        private readonly Func<CancellationToken, Task<Session>> sessionCreator;
+        private readonly IAsyncRetryHandler retryHandler;
+        private readonly ILogger logger;
```

Comment:
Make sure what can be shared here is shared.

Reply:
The readonly fields need to be here or else the constructors will complain during the build.

Reply:
Shared what can be shared.